### PR TITLE
Instruction stream tests first pass

### PR DESCRIPTION
**Pull request by StevenCostiou**
With @Ducasse we pair-programmed to understand the instruction stream and context classes.
Along the way, we write tests and fix little things (comments, dead code, etc.).

This is a first pass, there will be many others.

All comments are welcome.

### CODE REVIEW

**Comment by Ducasse** [comment]
Thanks Steven. We should continue because this is a super fun way to learn. 
I will ask marcus or pablo to review what we did and also get feedback on the renaming we propose.

File: src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	stream := InstructionStream on: compiledMethod.
+	client := InstructionClient new.
+	
+	"Answer the next byte.
+	For example for this bytecode 
+	<EB 00> superSend 
+	<D8> pop
+	 If the current byte is EB, the following byte is 00.
+	 It is not the following bytecode, that is D8"
+ 	stream pc: stream pc + 1.
```
Comment (guillep) [inline src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st:52]: I don't understand why the +1 here. It is assuming 00? But by reading the comment I would have thought your stream is at EB.

File: src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+]
+
+{ #category : #'tests - interpretation' }
+DebuggingInstructionStreamTest >> testInterpretNextInstruction [
+	| compiledMethod stream client currentPC |
+	compiledMethod := self class >> #method1.
+	stream := InstructionStream on: compiledMethod.
+	client := InstructionClient new.
+	currentPC := stream pc.
+	stream interpretNextInstructionFor: client.
```
Comment (guillep) [inline src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st:82]: This is to me a too generic name for the test. It really depends on what bytecode you are...

File: src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+{ #category : #'tests - structure' }
+DebuggingInstructionStreamTest >> testNextByte [
+
+	| compiledMethod stream client |
+	compiledMethod := self class >> #methodWithSuper.
+	stream := InstructionStream on: compiledMethod.
+	client := InstructionClient new.
+	stream interpretNextInstructionFor: client.
+	self assert: stream nextByte
```
Comment (guillep) [inline src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st:94]: What is the difference between nextByte and followingByte?

File: src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+{ #category : #'tests - structure' }
+DebuggingInstructionStreamTest >> testNextPc [
+
+	| compiledMethod stream bytecode |
+	compiledMethod := self class >> #method1.
+	stream := InstructionStream on: compiledMethod.
+	bytecode := stream firstByte.
+	self
+		assert: (stream nextPc: bytecode)
```
Comment (guillep) [inline src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st:119]: this method looks fishy. What happens if a same bytecode appears many times in the stream? Does it only take the first one? The name of the test is not really very intention revealing of what it is testing

File: src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	| compiledMethod stream client stream2 client2 |
+	compiledMethod := self class >> #methodWithSuper.
+	stream := InstructionStream on: compiledMethod.
+	client := InstructionClient new.
+	stream interpretNextInstructionFor: client.
+	
+	stream2 := InstructionStream on: compiledMethod.
+	client2 := InstructionClient new.
+	stream2 interpretNextInstructionFor: client2.
+	stream2 interpretNextInstructionFor: client2.
```
Comment (guillep) [inline src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st:136]: This line is repeated on purpose?

File: src/Debugging-Core/InstructionStream.extension.st [hunk 6]
```diff
@@ -24,12 +24,17 @@ InstructionStream >> debuggerMap [
 InstructionStream >> firstByte [
 	"Answer the first byte of the current bytecode."
 
-	^self method at: pc
+	^self method at: self method initialPC
```
Comment (guillep) [inline src/Debugging-Core/InstructionStream.extension.st:27]: This new implementation does not go well with the comment. The comment is supposed to return the first byte of the **current bytecode**. The new implementation returns the first byte of the method.

File: src/Debugging-Core/InstructionStream.extension.st [hunk 7]
```diff
@@ -253,37 +258,20 @@ InstructionStream >> interpretV3PlusClosureJumpIfCond [
 
 { #category : #'*Debugging-Core' }
 InstructionStream >> nextByte [
-	"Answer the next bytecode."
+	"Answer the next bytecode"
+	self flag: 'Should this method be renamed nextBytecode? Is it possible thta it returns only a byte?'.
 
 	^self method at: pc
```
Comment (guillep) [inline src/Debugging-Core/InstructionStream.extension.st:264]: What is the difference with followingByte? I was thinking this one advanced the stream, but it does not

File: src/Kernel-Tests/CompiledCodeTest.class.st [hunk 8]
```diff
@@ -30,6 +30,18 @@ CompiledCodeTest >> testAccessesRef [
...
+{ #category : #'tests - accessing' }
+CompiledCodeTest >> testBytecodeSizeForPc [
+
+	| method |
+	method := self class >> #method1.
+
+	self
+		assert: (method bytecodeSizeForPc: method initialPC)
+		equals:
+		(method encoderClass bytecodeSize: (method at: method initialPC))
```
Comment (guillep) [inline src/Kernel-Tests/CompiledCodeTest.class.st:42]: Again, this test really depends on the input, and different inputs do different things. Maybe they need different tests...

**Review by guillep — COMMENTED** [review]
I have some comments, there is one that looks like a bug, so be careful...

File: src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st (hunk 5 above)
Comment (Ducasse) [reply src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st:136]: Yes. We eat two.

File: src/Debugging-Core/InstructionStream.extension.st (hunk 7 above)
Comment (Ducasse) [reply src/Debugging-Core/InstructionStream.extension.st:264]: Indeed this is what we faced. nextByte is the nextByteCode while followingByte is the following byte of the byte code

**Comment by Ducasse** [comment]
Thanks we will check your comment. I wish to have the dumbledoor stuff to remember :).

File: src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st (hunk 3 above)
Comment (Ducasse) [reply src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st:94]: If I remember one is a bytecode and the other a byte so we found that super confusing.

File: src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st (hunk 1 above)
Comment (Ducasse) [reply src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st:52]: Here we wanted to show is that if we have a bytecode with two bytes, the followingByte (not bytecode) is the one following the one pointed by the current PC
So we point at EB and followingByte should be 00. 
May be the test is wrong this is what we want to test.

**Comment by StevenCostiou** [comment]
Hi, thanks for the reviews!
Sorry I am now on holidays even if I keep an eye from time to time.
After I come back on March 8th I will come back to that.

File: src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st (hunk 2 above)
Comment (Ducasse) [reply src/Debugging-Core-Tests/DebuggingInstructionStreamTest.class.st:82]: agreed

File: src/Debugging-Core/InstructionStream.extension.st (hunk 6 above)
Comment (Ducasse) [reply src/Debugging-Core/InstructionStream.extension.st:27]: agreed!

File: src/Kernel-Tests/CompiledCodeTest.class.st (hunk 8 above)
Comment (Ducasse) [reply src/Kernel-Tests/CompiledCodeTest.class.st:42]: Probably we were started to see.

File: src/Debugging-Core/InstructionStream.extension.st (hunk 7 above)
Comment (Ducasse) [reply src/Debugging-Core/InstructionStream.extension.st:264]: to me the current definition is super bad

```
followingByte
	"Answer the next bytecode."

	^self method at: pc + 1
```

**Comment by Ducasse** [comment]
Stevn I will close it and I rescued it because I still would to continue.

**Comment by guillep** [comment]
This was already redone and integrated in Pharo10

**Pull request closed without merging**